source/Calamari.Aws/Integration/S3/S3ObjectExtensions.cs

```diff
@@ -4,16 +4,101 @@
 using Amazon.S3.Model;
 ﻿using System;
 using System.Security.Cryptography;
- using Calamari.Integration.FileSystem;
+using Calamari.Integration.FileSystem;
 using Calamari.Util;
 using Octopus.CoreUtilities;
 using Octopus.CoreUtilities.Extensions;
- using Tag = Amazon.S3.Model.Tag;
+using Tag = Amazon.S3.Model.Tag;
 
 namespace Calamari.Aws.Integration.S3
 {
     public static class S3ObjectExtensions
     {
+        //Special headers as per AWS docs - https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html
+        private static readonly IDictionary<string, Func<HeadersCollection, string>> SupportedSpecialHeaders =
+            new Dictionary<string, Func<HeadersCollection, string>>(StringComparer.InvariantCultureIgnoreCase)
+                .WithHeaderFrom("Cache-Control", headers => headers.CacheControl)
+                .WithHeaderFrom("Content-Disposition", headers => headers.ContentDisposition)
+                .WithHeaderFrom("Content-Encoding", headers => headers.ContentEncoding)
+                .WithHeaderFrom("Expires")
+                .WithHeaderFrom("x-amz-website​-redirect-location")
+                .WithHeaderFrom("x-amz-object-lock-mode");
+
+        private static T WithHeaderFrom<T>(this T values, string key) where T : IDictionary<string, Func<HeadersCollection, string>>
+        {
+            values.Add(key, (headers) => headers[key]);
+            return values;
+        }
+
+        private static T WithHeaderFrom<T>(this T values, string key, Func<HeadersCollection, string> accessor) where T : IDictionary<string, Func<HeadersCollection, string>>
+        {
+            values.Add(key, accessor);
+            return values;
+        }
+
+        public static IEnumerable<string> GetKnownSpecialHeaderKeys()
+        {
+            return SupportedSpecialHeaders.Keys;
+        }
+
+        public static bool MetadataEq(IDictionary<string, string> next, IDictionary<string, string> current)
+        {
+            var allKeys = next.Keys.Union(current.Keys).Distinct();
+            var keysInBoth = new List<string>();
```

Comment:
Nice! Now this is where I get all functional! :)

`keysInBoth = allKeys.Where(key => next.ContainsKey(key) && current.ContainsKey(key)).ToArray();`
`missingKeys = allKeys.Except(keysInBoth).ToArray()`
`differentValues = keysInBoth.Where(key => ...).ToArray()`
`return missingKeys.Count == 0 && differentValues == 0;`

:)

Reply:
Ah I think I see where this is going, getting the differing values gives us an opportunity to also add that information to the logs to give some indication at least of what has actually changed. Is that what you were thinking?

Reply:
Well, it's a certain way of structuring these kinds of functions into decomposed queries. The knock on effects are:
1. The code becomes more readable since each query means something sensible, and the final result is usually an aggregate of those things
2. If it's easier to reason about each query, it's easier to maintain the code later
3. You can use this information (if you really want to) for logging/tracing/debugging.

It's just a habit I've built after falling in love with LINQ and idiomatic functional programming in C#. I find future self thanking my past self when I take the time to think this way.